2199/2/versus.rb

```diff
@@ -0,0 +1,83 @@
+require 'pry'
+require 'yaml'
+require 'russian_obscenity'
+require 'terminal-table'
+require 'optparse'
+require 'russian'
+
+ALIASES = YAML.load_file('alias.yml')
+rappers = {}
+# :reek:TooManyInstanceVariables
+# Participant container
+class Participant
+  attr_reader :name, :bad_words, :battle_count
+  def initialize(name)
+    @name = name
+    @battle_count = 0
+    @bad_words = 0
+    @words_count = 0
+    @rounds_count = 0
+  end
+
+  # :reek:NilCheck
+  def self.name_from_file(filename)
+    name = filename.split(/ против | vs | VS /).first.split('/').last.strip
+    ALIASES.find do |_, aliases|
+      aliases.include?(name)
+    end&.first || name
+  end
+
+  def filthy(filename)
+    @battle_count += 1
+    text = File.read(filename)
+    words = text.scan(/[а-яА-ЯёЁ*]+/)
+    words.delete('***')
+    words_and_round_counts(words, text)
```

Comment:
Right now `Participant` class look like a state storage and a number of procedures that change the state. For example, you call `#filthy` method, then this method calls the `#words_and_round_counts` and it calls `#bad_words_find`.  This code can easily be written as one big function that changes state line by line. What you do is cheating with reek.

Also, this `Participant` class do many jobs: it represents a participant, it calculates statistics for battles for participants, it also represents the output row.

With OOP we can do better ;-). First thing is we should avoid calling methods one by one in order to tell them what to do.

Let me show you how I would think if I were designing this code.

1. Ok, I need to show top N participants by usages of bad words. Let it be a class called `Participants` which will return participants.
```rb
class Participants
  def initialize()
  end

  def top_by_bad_words
    # ...
  end
end
```

2. And let there be a code that gets participants and output them:
```rb
participants = Participants.new
puts Terminal::Table.new(rows: participants.top_by_bad_words(top_n).map do |participant|
    # ...
  end
)
```
Don't know how to call this code yet.

3. Let's try to expand `Participant#top_by_bad_words`:
```rb
def top_by_bad_words(limit)
  sorted_by_bad_words.first(limit)
end
```

4. Go further. I write kind of pseudo-code first, w/o thinking about implementation of methods I call. I just pretent they exist and later I will write them.
```rb
def sorted_by_bad_words
  participants.sort_by # ...
end
```

5. Hm... `#participants`:
```rb
def participants
  name_and_rounds.map { |name, rounds| Participant.new(name, rounds) }
end
```
I just pretent there is `name_and_rounds` method which returns hash which maps names to array of strings. And here I introduced a `Participant` class which also is not implemented yet. But my idea is to pass name and rounds to `Participant` and then it will have a methods that return need statistics (number of bad words, bad words per round, etc.).

I will stop here. @LikaLex, do you follow my thoughts? Let's chnage our code the way classes are small, nice and have s single responsibility.